File: 57-insert-interval/57-insert-interval.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> list = new ArrayList<>();
        
        for(int[] curr : intervals) {
            if(newInterval == null || curr[1] < newInterval[0]) {
                list.add(curr);
            } else if(curr[0] > newInterval[1]) {
                list.add(newInterval);
                list.add(curr);
                newInterval = null;
            } else {
                newInterval[0] = Math.min(curr[0], newInterval[0]);
                newInterval[1] = Math.max(curr[1], newInterval[1]);
            }
        }
        if(newInterval != null) list.add(newInterval);
        return list.toArray(new int[list.size()][2]);
    }
}

File: 57-insert-interval/57-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        intervals.sort(key=lambda x :x[0])
        
        res = []
        for interval in intervals:
            if not newInterval or newInterval[0] > interval[1]:
                res.append(interval)
            elif newInterval[1] < interval[0]:
                res.append(newInterval)
                res.append(interval)
                newInterval = None
            else:
                newInterval[0] = min(newInterval[0], interval[0])
                newInterval[1] = max(newInterval[1], interval[1])
        if newInterval:
            res.append(newInterval)
        return res